File: mobile/android/app/src/main/kotlin/Views/Chats/ChatsView.kt
package net.theavalanche.app

import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.core.animateDpAsState
import androidx.compose.animation.core.LinearOutSlowInEasing
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.indication
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Create
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateMapOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

// ---------------------------------------------------------------------------
// ChatsView
//
// Mirrors iOS Sources/Views/Chats/ChatsView.swift (docs/37-chat-organization):
// no top app bar — a custom header hosts the per-account tab strip (only when
// the user has >1 account; Signal-baseline progressive disclosure) with the
// compose button on the trailing side. Selection on the strip is an avBrand
// ring around the account's avatar (iOS-style), animated between tabs. The
// conversation list sits on `card` (chrome stays `paper`), rows have a
// press highlight, and previews are collapsed to their first line.
//
// Navigation callbacks replace NavigationStack — the central NavGraph wires
// them. Settings moved to the bottom NavigationBar (MainTabView).
// ---------------------------------------------------------------------------

@Composable
fun ChatsView(
    viewModel: AppViewModel,
    onOpenConversation: (Conversation) -> Unit = {},
    onOpenAccounts: () -> Unit = {},
    onOpenCompose: () -> Unit = {},
) {
    val conversations by viewModel.conversations.collectAsState()
    val conversationsLoaded by viewModel.conversationsLoaded.collectAsState()
    val navigateToConversation by viewModel.navigateToConversation.collectAsState()
    val accounts by viewModel.accounts.collectAsState()

    // Mirror iOS onChange(of: appState.navigateToConversation)
    LaunchedEffect(navigateToConversation) {
        val conv = navigateToConversation ?: return@LaunchedEffect
        viewModel.setNavigateToConversation(null)
        onOpenConversation(conv)
    }

    // Whether the per-account tab strip is shown at all. Progressive
    // disclosure: a single-account user sees a plain unified inbox (docs/37).
    val showsAccountTabs = accounts.size > 1

    // The account (identity) whose conversations are shown when the strip is
    // visible. Held in the ViewModel (survives bottom-tab switches — local
    // state would reset and flash an unfiltered frame); the effective value is
    // derived *synchronously* so no frame ever renders with a stale/absent
    // selection: fall back to the first account if unset or vanished.
    val rawSelectedAccountTab by viewModel.selectedAccountTab.collectAsState()
    val selectedAccountTab = rawSelectedAccountTab
        ?.takeIf { id -> accounts.any { it.id == id } }
        ?: accounts.firstOrNull()?.id

    val sortedConversations = remember(conversations) {
        conversations.sortedByDescending { it.lastMessageDate?.time ?: Long.MIN_VALUE }
    }
    val visibleConversations = remember(sortedConversations, showsAccountTabs, selectedAccountTab) {
        val selected = selectedAccountTab
        if (showsAccountTabs && selected != null) {
            sortedConversations.filter { it.accountId == selected }
        } else {
            sortedConversations
        }
    }

    // TODO(opus): hasRecoveryKey — check via Rust core once the FFI method exists.
    val hasRecoveryKey = false

    Scaffold(
        topBar = {
            ChatsHeader(
                accounts = accounts,
                showsAccountTabs = showsAccountTabs,
                selectedAccountTab = selectedAccountTab,
                onSelectAccountTab = { viewModel.setSelectedAccountTab(it) },
                onOpenCompose = onOpenCompose,
            )
        },
        containerColor = LocalAvalancheColors.current.paper,
    ) { innerPadding ->
        Box(
            modifier = Modifier
                .fillMaxSize()
                .padding(innerPadding),
        ) {
            if (sortedConversations.isEmpty() && !conversationsLoaded) {
                // Initial load still in flight — show a spinner instead of the
                // empty state, so "No conversations yet" doesn't flash on launch.
                Box(
                    modifier = Modifier
                        .fillMaxSize()
                        .background(LocalAvalancheColors.current.card),
                    contentAlignment = Alignment.Center,
                ) {
                    CircularProgressIndicator(color = LocalAvalancheColors.current.muted)
                }
            } else if (visibleConversations.isEmpty()) {
                // Mirrors iOS ContentUnavailableView
                Column(
                    modifier = Modifier
                        .fillMaxSize()
                        .background(LocalAvalancheColors.current.card)
                        .padding(32.dp),
                    horizontalAlignment = Alignment.CenterHorizontally,
                    verticalArrangement = Arrangement.Center,
                ) {
                    Icon(
                        imageVector = Icons.Filled.Create,
                        contentDescription = null,
                        tint = LocalAvalancheColors.current.muted,
                        modifier = Modifier.size(48.dp),
                    )
                    Spacer(Modifier.size(16.dp))
                    Text(
                        text = "No conversations yet",
                        color = LocalAvalancheColors.current.ink,
                        fontSize = 18.sp,
                        fontWeight = FontWeight.SemiBold,
                    )
                    Spacer(Modifier.size(8.dp))
                    Text(
                        text = "Messages from all your servers will appear here.",
                        color = LocalAvalancheColors.current.muted,
                        fontSize = 14.sp,
                    )
                }
            } else {
                Column(
                    modifier = Modifier
                        .fillMaxSize()
                        .background(LocalAvalancheColors.current.card),
                ) {
                    // Recovery key banner (mirrors iOS placement below the header)
                    if (!hasRecoveryKey) {
                        RecoveryKeyBanner()
                        HorizontalDivider(color = LocalAvalancheColors.current.divider)
                    }

                    LazyColumn(
                        modifier = Modifier.fillMaxSize(),
                        // Content scrolls behind the floating bottom bar; the
                        // inset lets the last row scroll clear of it.
                        contentPadding = PaddingValues(bottom = FloatingBarContentPadding),
                    ) {
                        items(
                            items = visibleConversations,
                            key = { it.id },
                        ) { conversation ->
                            val account = accounts.firstOrNull { it.id == conversation.accountId }
                            val unreadCount = viewModel.unreadCount(conversation)
                            val recipientDid = conversation.recipientDid
                            val isBot = !conversation.isGroup &&
                                recipientDid != null &&
                                viewModel.isBot(recipientDid, conversation.accountId)
                            ConversationRow(
                                conversation = conversation,
                                account = account,
                                accounts = accounts,
                                unreadCount = unreadCount,
                                isBotConversation = isBot,
                                previewText = conversationPreviewText(viewModel, conversation),
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .chatRowPressHighlight { onOpenConversation(conversation) }
                                    .padding(horizontal = 16.dp, vertical = 10.dp),
                            )
                            HorizontalDivider(
                                color = LocalAvalancheColors.current.divider.copy(alpha = 0.5f),
                                modifier = Modifier.padding(start = 76.dp),
                            )
                        }
                    }
                }
            }
        }
    }
}

// ---------------------------------------------------------------------------
// Header: account tab strip + compose
// ---------------------------------------------------------------------------

/**
 * Custom top header replacing the old top app bar: the per-account tab strip
 * (only with >1 account) on the leading side, compose on the trailing side.
 * Compose is deliberately *not* part of the tab strip (docs/37).
 */
@Composable
private fun ChatsHeader(
    accounts: List<Account>,
    showsAccountTabs: Boolean,
    selectedAccountTab: String?,
    onSelectAccountTab: (String) -> Unit,
    onOpenCompose: () -> Unit,
) {
    // background *then* statusBarsPadding: paper paints up behind the status
    // bar, content starts below it. (No outer Scaffold supplies this inset —
    // the floating-bar layout owns the whole window.)
    Column(
        modifier = Modifier
            .background(LocalAvalancheColors.current.paper)
            .statusBarsPadding(),
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 16.dp)
                // No bottom padding: the tab strip's sliding underline sits at
                // the bottom of this row, flush with the header's bottom edge.
                .padding(top = 4.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            if (showsAccountTabs) {
                // Material-idiom selection: a brand underline that slides
                // between tabs (Android-y), instead of iOS's avatar ring.
                // Tabs are fixed-width (68dp + 8dp spacing), so the indicator
                // offset is a simple animated multiple of the slot width.
                val selectedIndex = accounts.indexOfFirst { it.id == selectedAccountTab }
                val slotWidth = 68.dp + 8.dp
                // Modern M3 "primary" indicator: a 3dp pill with rounded top
                // corners, sized to the selected tab's *label text* and
                // centered under it. Label widths are measured per tab (they
                // vary by name) and reported up via onLabelWidth.
                val labelWidths = remember { mutableStateMapOf<String, Dp>() }
                Box(
                    modifier = Modifier
                        .weight(1f)
                        .horizontalScroll(rememberScrollState()),
                ) {
                    Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                        accounts.forEach { account ->
                            AccountTab(
                                account = account,
                                selected = selectedAccountTab == account.id,
                                onClick = { onSelectAccountTab(account.id) },
                                onLabelWidth = { labelWidths[account.id] = it },
                            )
                        }
                    }
                    // The indicator (and its animated values) only enters
                    // composition once the selected label's width has been
                    // measured — animateDpAsState initializes *at* its first
                    // target, so the first render lands at final size/position
                    // with no settle wiggle; later tab switches animate.
                    val measuredWidth = selectedAccountTab?.let { labelWidths[it] }
                    if (selectedIndex >= 0 && measuredWidth != null) {
                        val indicatorWidth by animateDpAsState(
                            targetValue = measuredWidth,
                            animationSpec = tween(durationMillis = 130, easing = LinearOutSlowInEasing),
                            label = "accountTabIndicatorWidth",
                        )
                        val indicatorOffset by animateDpAsState(
                            targetValue = slotWidth * selectedIndex +
                                (68.dp - indicatorWidth) / 2,
                            animationSpec = tween(durationMillis = 130, easing = LinearOutSlowInEasing),
                            label = "accountTabIndicator",
                        )
                        Box(
                            modifier = Modifier
                                .align(Alignment.BottomStart)
                                .padding(start = indicatorOffset)
                                .width(indicatorWidth)
                                .height(3.dp)
                                .background(
                                    color = LocalAvalancheColors.current.brand,
                                    shape = RoundedCornerShape(
                                        topStart = 3.dp,
                                        topEnd = 3.dp,
                                    ),
                                ),
                        )
                    }
                }
            } else {
                Spacer(modifier = Modifier.weight(1f))
            }
            IconButton(onClick = onOpenCompose) {
                Icon(
                    imageVector = Icons.Filled.Create,
                    contentDescription = "New message",
                    tint = LocalAvalancheColors.current.brand,
                )
            }
        }
        // No divider/elevation below the header — the paper→card tint change
        // is the separation; a hairline under the indicator reads as a thick
        // double border.
    }
}

/**
 * One account tab: the identity's avatar (monogram fallback) above its name.
 * Selection is shown Material-style — the sliding brand underline drawn by the
 * strip (not iOS's avatar ring); the selected avatar is full-color and the
 * others dimmed, animated with a short no-overshoot tween. Auto-derived so a
 * per-account tab is meaningful with zero config; user-created/topic tabs will
 * carry a chosen symbol instead (tab organizer, deferred — docs/37).
 */
@Composable
private fun AccountTab(
    account: Account,
    selected: Boolean,
    onClick: () -> Unit,
    // Reports the measured width of the label text, so the strip can size its
    // sliding indicator pill to match.
    onLabelWidth: (Dp) -> Unit = {},
) {
    val density = LocalDensity.current
    val interaction = remember { MutableInteractionSource() }
    val pressed by interaction.collectIsPressedAsState()
    val pressScale by animateFloatAsState(
        targetValue = if (pressed) 0.93f else 1f,
        animationSpec = tween(durationMillis = 120),
        label = "accountTabPress",
    )

    val avatarAlpha by animateFloatAsState(
        targetValue = if (selected) 1f else 0.45f,
        animationSpec = tween(durationMillis = 130, easing = LinearOutSlowInEasing),
        label = "accountTabAlpha",
    )
    val avatarScale by animateFloatAsState(
        targetValue = if (selected) 1f else 0.9f,
        animationSpec = tween(durationMillis = 130, easing = LinearOutSlowInEasing),
        label = "accountTabScale",
    )

    val label = accountTabLabel(account)

    Column(
        modifier = Modifier
            .scale(pressScale)
            .clip(RoundedCornerShape(14.dp))
            .background(
                if (pressed) LocalAvalancheColors.current.ink.copy(alpha = 0.08f)
                else Color.Transparent,
            )
            .clickable(
                interactionSource = interaction,
                indication = null,
                onClick = onClick,
            )
            .width(68.dp)
            // Bottom padding leaves room for the strip's sliding underline.
            .padding(top = 2.dp, bottom = 6.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(5.dp),
    ) {
        Box(
            modifier = Modifier
                .size(30.dp)
                .alpha(avatarAlpha)
                .scale(avatarScale)
                .background(
                    color = LocalAvalancheColors.current.card,
                    shape = CircleShape,
                )
                .clip(CircleShape),
            contentAlignment = Alignment.Center,
        ) {
            Text(
                text = label.firstOrNull()?.uppercaseChar()?.toString() ?: "?",
                fontSize = 14.sp,
                fontWeight = FontWeight.SemiBold,
                color = LocalAvalancheColors.current.brand,
            )
        }
        Text(
            text = label,
            fontSize = 11.sp,
            fontWeight = if (selected) FontWeight.SemiBold else FontWeight.Normal,
            // Selected label matches the indicator (brand) — M3 primary-tab
            // idiom. TODO(design): consider a brighter accent if plum reads
            // too quiet here.
            color = if (selected) LocalAvalancheColors.current.brand else LocalAvalancheColors.current.muted,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
            onTextLayout = { layout ->
                onLabelWidth(with(density) { layout.size.width.toDp() })
            },
        )
    }
}

/**
 * A short label for an account tab: the identity's display name, falling back
 * to its first server's compact host when the name is unset.
 */
private fun accountTabLabel(account: Account): String {
    if (account.displayName.isNotEmpty()) return account.displayName
    return account.servers.firstOrNull()?.let { it.url.host ?: it.name } ?: "Account"
}

// ---------------------------------------------------------------------------
// Shared helpers (also used by ConversationSearchView)
// ---------------------------------------------------------------------------

/**
 * Press feedback for a chat row: a subtle ink highlight while the finger is
 * down (over the row's `card` background). Mirrors iOS ChatRowButtonStyle.
 */
@Composable
fun Modifier.chatRowPressHighlight(onClick: () -> Unit): Modifier {
    val interaction = remember { MutableInteractionSource() }
    val pressed by interaction.collectIsPressedAsState()
    return this
        .background(
            if (pressed) LocalAvalancheColors.current.ink.copy(alpha = 0.06f)
            else Color.Transparent,
        )
        .clickable(
            interactionSource = interaction,
            indication = null,
            onClick = onClick,
        )
}

/**
 * Preview text for the latest message — mirrors iOS ConversationRow.previewText.
 * For a group system event we render it reactively (resolving DIDs to names at
 * display time). Normal messages compose the content decoration (📷/📎/👤) with
 * the body and the "You:"/sender prefix. Collapsed to the first line — bodies
 * can contain hard line breaks and the chat list must never render past line
 * one (docs/37).
 *
 * Computed live (not `remember`d) so the snapshot reads of `displayNameCache`
 * inside resolvedName/groupEventText are tracked: when a name resolves async,
 * the row recomposes and "Unknown" becomes the real name.
 */
@Composable
fun conversationPreviewText(viewModel: AppViewModel, conversation: Conversation): String? {
    if (conversation.lastMessageKind > 0) {
        val msg = Message(
            id = conversation.id,
            conversationId = conversation.id,
            senderAccountId = conversation.lastMessageSenderDid ?: "",
            body = conversation.lastMessage ?: "",
            sentAtMs = 0L,
            readAtMs = null,
            deliveryStatus = DeliveryStatus.SENT,
            kind = conversation.lastMessageKind,
            metadata = conversation.lastMessageMetadata,
        )
        return firstLinePreview(viewModel.groupEventText(msg, conversation.accountId))
    }
    // Compose the content decoration (📷/📎/👤) with the body (docs/35): a
    // caption shows "📷 caption", a caption-less content message shows "📷
    // Photo", and plain text shows just the body. Kept in sync between live
    // and post-restart previews via `lastMessagePreview`.
    val rawBody = firstLinePreview(conversation.lastMessage ?: "")
    val deco = lastMessagePreviewDecoration(conversation.lastMessagePreview)
    val body: String = when {
        deco != null -> if (rawBody.isEmpty()) "${deco.first} ${deco.second}" else "${deco.first} $rawBody"
        rawBody.isNotEmpty() -> rawBody
        else -> return null
    }
    val sender = conversation.lastMessageSenderDid
    return when {
        sender != null && sender == conversation.accountId -> "You: $body"
        conversation.isGroup && !sender.isNullOrEmpty() ->
            "${viewModel.resolvedName(sender, conversation.accountId)}: $body"
        else -> body
    }
}

// ---------------------------------------------------------------------------
// Preview
// ---------------------------------------------------------------------------

@Preview(showBackground = true)
@Composable
private fun ChatsViewEmptyPreview() {
    AvalancheTheme {
        Box(
            modifier = Modifier
                .fillMaxSize()
                .background(LocalAvalancheColors.current.card),
            contentAlignment = Alignment.Center,
        ) {
            Column(
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                Text(
                    text = "No conversations yet",
                    color = LocalAvalancheColors.current.ink,
                    fontWeight = FontWeight.SemiBold,
                    fontSize = 18.sp,
                )
                Text(
                    text = "Messages from all your servers will appear here.",
                    color = LocalAvalancheColors.current.muted,
                    fontSize = 14.sp,
                )
            }
        }
    }
}

File: mobile/ios/Actnet/Sources/Views/Chats/ChatsView.swift
import SwiftUI

struct ChatsView: View {
    @EnvironmentObject var appState: AppState
    @State private var showCompose = false
    @State private var navigationPath = NavigationPath()
    /// Namespace for the selection ring so it animates *between* avatars
    /// (a single moving indicator) rather than fading in/out per tab.
    @Namespace private var tabNamespace

    /// The account (identity) whose conversations are shown, when the user has
    /// more than one account and the tab strip is visible. Selection lives in
    /// AppState (survives navigation); the *effective* value is derived
    /// synchronously — explicit choice if it names a live account, else the
    /// first account — so the first render is already filtered (no unfiltered
    /// flash at cold launch). See `docs/37-chat-organization.md`.
    private var selectedAccountTab: String? {
        guard showsAccountTabs else { return nil }
        if let sel = appState.selectedChatsAccountTab,
           appState.accounts.contains(where: { $0.id == sel }) {
            return sel
        }
        return appState.accounts.first?.id
    }

    /// Whether the per-account tab strip is shown at all. Progressive disclosure:
    /// a single-account user sees a plain unified inbox (Signal baseline, `37`).
    private var showsAccountTabs: Bool {
        appState.accounts.count > 1
    }

    /// Conversations sorted by most recent message first.
    private var sortedConversations: [Conversation] {
        appState.conversations.sorted { a, b in
            (a.lastMessageDate ?? .distantPast) > (b.lastMessageDate ?? .distantPast)
        }
    }

    /// The conversations for the active view: filtered to the selected account
    /// tab when the strip is shown, otherwise the full unified list.
    private var visibleConversations: [Conversation] {
        guard showsAccountTabs, let selected = selectedAccountTab else {
            return sortedConversations
        }
        return sortedConversations.filter { $0.accountId == selected }
    }

    var body: some View {
        NavigationStack(path: $navigationPath) {
            VStack(spacing: 0) {
                header
                if !hasRecoveryKey {
                    RecoveryKeyBanner()
                }
                content
            }
            .background(Color.avPaper.ignoresSafeArea())
            .toolbar(.hidden, for: .navigationBar)
            .navigationDestination(for: Conversation.self) { conversation in
                ConversationView(conversation: conversation)
            }
            .sheet(isPresented: $showCompose) {
                ComposeMessageView()
            }
            .onChange(of: appState.navigateToConversation) {
                guard let conv = appState.navigateToConversation else { return }
                appState.navigateToConversation = nil
                // Replace the whole path in one atomic update: root → conversation.
                // The previous reset-to-empty + `DispatchQueue.main.async` append
                // was two mutations across runloops; on the deep-link path it
                // raced with the tab switch and the still-dismissing ProjectWebView
                // sheet, landing on a blank pushed view. A single assignment lands
                // cleanly at root → conversation with no intermediate empty state.
                navigationPath = NavigationPath([conv])
            }
        }
    }

    // MARK: Header (account tabs + compose)

    /// Custom top header replacing the old nav bar: the per-account tab strip
    /// (only with >1 account) on the leading side, compose on the trailing side.
    /// Compose is deliberately *not* part of the tab strip (`37`).
    private var header: some View {
        HStack(alignment: .center, spacing: 8) {
            if showsAccountTabs {
                accountTabStrip
            } else {
                Spacer()
            }
            composeButton
        }
        .padding(.horizontal, 16)
        .padding(.top, 4)
        .padding(.bottom, 6)
        .background(Color.avPaper)
        .overlay(alignment: .bottom) {
            Rectangle()
                .fill(Color.avInk.opacity(0.08))
                .frame(height: 0.5)
        }
    }

    private var accountTabStrip: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 8) {
                ForEach(appState.accounts) { account in
                    accountTab(account)
                }
            }
            // A single ring that follows the selected avatar's frame. Because
            // the avatars are the geometry *sources* and this is the sole
            // non-source follower, the slide is one continuous move — so the
            // duration below is the real speed knob (no insert/remove snap).
            .overlay {
                if let selected = selectedAccountTab {
                    Circle()
                        .strokeBorder(Color.avBrand, lineWidth: 2)
                        .frame(width: 36, height: 36)
                        .matchedGeometryEffect(id: selected, in: tabNamespace, isSource: false)
                        .allowsHitTesting(false)
                }
            }
            .animation(.easeOut(duration: 0.13), value: appState.selectedChatsAccountTab)
        }
    }

    private func accountTab(_ account: Account) -> some View {
        let selected = selectedAccountTab == account.id
        return Button {
            appState.selectedChatsAccountTab = account.id
        } label: {
            VStack(spacing: 5) {
                accountIcon(account, selected: selected)
                Text(accountLabel(account))
                    .font(.caption)
                    .fontWeight(selected ? .semibold : .regular)
                    .foregroundStyle(selected ? Color.avInk : Color.avMuted)
                    .lineLimit(1)
            }
            .frame(width: 68)
            .padding(.vertical, 2)
        }
        .buttonStyle(AccountTabButtonStyle())
    }

    /// The tab's icon: the identity's own avatar, falling back to a monogram
    /// circle. Selection is shown iOS-style — an avBrand ring that slides
    /// between avatars (via `matchedGeometryEffect`), selected avatar full and
    /// slightly enlarged, others dimmed (no underline). Auto-derived so a
    /// per-account tab is meaningful with zero config; user-created/topic tabs
    /// will carry a chosen symbol instead (tab organizer, deferred — `docs/37`).
    private func accountIcon(_ account: Account, selected: Bool) -> some View {
        avatarThumbnail(account)
            .frame(width: 30, height: 30)
            .clipShape(Circle())
            .opacity(selected ? 1 : 0.45)
            .scaleEffect(selected ? 1 : 0.9)
            .padding(3)
            // Publish this avatar's (36×36 padded) frame as a ring geometry
            // source; the single follower ring in the strip overlay matches
            // whichever one is selected.
            .matchedGeometryEffect(id: account.id, in: tabNamespace, isSource: true)
            // Unread badge — sum across the tab's conversations, same
            // notification/white style as the row badge, riding the avatar's
            // top-trailing corner.
            .overlay(alignment: .topTrailing) {
                let unread = unreadCount(for: account)
                if unread > 0 {
                    Text("\(unread)")
                        .font(.system(size: 9, weight: .bold))
                        .foregroundStyle(.white)
                        .padding(.horizontal, 4)
                        .padding(.vertical, 1)
                        .background(Color.avNotification, in: Capsule())
                        .offset(x: 9, y: -4)
                        .allowsHitTesting(false)
                }
            }
    }

    /// Total unread messages across this account's conversations — rendered
    /// as a notification badge on the tab avatar (matches the row badge style).
    private func unreadCount(for account: Account) -> Int {
        appState.conversations
            .filter { $0.accountId == account.id }
            .reduce(0) { $0 + appState.unreadCount(for: $1) }
    }

    @ViewBuilder
    private func avatarThumbnail(_ account: Account) -> some View {
        if let data = account.avatarData, let image = UIImage(data: data) {
            Image(uiImage: image)
                .resizable()
                .scaledToFill()
        } else {
            Circle()
                .fill(Color.avCard)
                .overlay {
                    Text(accountLabel(account).prefix(1).uppercased())
                        .font(.subheadline)
                        .fontWeight(.semibold)
                        .foregroundStyle(Color.avBrand)
                }
        }
    }

    private var composeButton: some View {
        Button {
            showCompose = true
        } label: {
            Image(systemName: "square.and.pencil")
                .font(.title3)
                .foregroundStyle(Color.avBrand)
                .frame(width: 44, height: 44)
        }
        .accessibilityLabel("New message")
    }

    // MARK: Conversation list

    @ViewBuilder
    private var content: some View {
        if appState.conversations.isEmpty && !appState.conversationsLoaded {
            // Initial load still in flight — show a spinner instead of the
            // empty state, so "No conversations yet" doesn't flash on launch.
            ProgressView()
                .frame(maxWidth: .infinity, maxHeight: .infinity)
                .background(Color.avCard)
        } else if visibleConversations.isEmpty {
            ContentUnavailableView(
                "No conversations yet",
                systemImage: "message",
                description: Text("Messages from all your servers will appear here.")
            )
            .background(Color.avCard)
        } else {
            conversationList
        }
    }

    private var conversationList: some View {
        List {
            ForEach(visibleConversations) { conversation in
                Button {
                    navigationPath.append(conversation)
                } label: {
                    ConversationRow(conversation: conversation, account: accountFor(conversation))
                }
                .buttonStyle(ChatRowButtonStyle())
                .listRowInsets(EdgeInsets())
                .listRowBackground(Color.avCard)
            }
        }
        .listStyle(.plain)
        .scrollContentBackground(.hidden)
        .background(Color.avCard)
    }

    // MARK: Helpers

    /// A short label for an account tab: the identity's display name, falling
    /// back to its first server's compact host when the name is unset.
    private func accountLabel(_ account: Account) -> String {
        if !account.displayName.isEmpty { return account.displayName }
        return account.servers.first?.displayHost ?? "Account"
    }

    private var hasRecoveryKey: Bool {
        // TODO: Check via Rust core
        false
    }

    private func accountFor(_ conversation: Conversation) -> Account? {
        appState.accounts.first { $0.id == conversation.accountId }
    }
}

/// Press feedback for a chat row: a full-bleed backing highlight while the
/// finger is down (over the row's `avCard`), released on lift. Rows are Buttons
/// (not NavigationLinks) so this press state is controllable — and so they get
/// no auto-disclosure chevron, matching a Signal/Messages-style chat list.
struct ChatRowButtonStyle: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .padding(.horizontal, 16)
            .padding(.vertical, 6)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(configuration.isPressed ? Color.avInk.opacity(0.06) : Color.clear)
            .contentShape(Rectangle())
    }
}

/// Press feedback for an account tab: a subtle backing highlight + slight
/// shrink while the finger is down, released on lift. Immediate (short ease),
/// independent of the springy selection animation.
private struct AccountTabButtonStyle: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .background {
                RoundedRectangle(cornerRadius: 14, style: .continuous)
                    .fill(Color.avInk.opacity(configuration.isPressed ? 0.08 : 0))
            }
            .scaleEffect(configuration.isPressed ? 0.93 : 1)
            .animation(.easeOut(duration: 0.12), value: configuration.isPressed)
    }
}

#if DEBUG
@MainActor
private func previewChatsState(accounts: [Account]) -> AppState {
    let state = AppState.preview(accounts: accounts)
    state.conversationsLoaded = true
    let now = Date()
    func conv(_ id: String, _ title: String, _ account: Account, _ body: String,
              _ ago: TimeInterval, group: Bool = false) -> Conversation {
        Conversation(
            id: id, title: title, accountId: account.id,
            serverUrl: account.servers.first?.url.absoluteString ?? "",
            lastMessage: body, lastMessageDate: now.addingTimeInterval(-ago), isGroup: group
        )
    }
    let real = accounts[0]
    var convos = [
        conv("c1", "Canvass NW", real, "See you at 5", 120, group: true),
        conv("c2", "Alice", real, "sounds good", 900),
        conv("c3", "Phone Bank", real, "who's in tonight?", 3600, group: true),
    ]
    if accounts.count > 1 {
        let persona = accounts[1]
        convos += [
            conv("c4", "Bob", persona, "hey there", 7200),
            conv("c5", "Reading Group", persona, "chapter 4 thoughts?", 90_000, group: true),
        ]
    }
    state.conversations = convos
    state.unreadCounts = ["c1": 3, "c3": 1, "c4": 2]
    return state
}

private func previewAccount(_ id: String, _ name: String, _ host: String) -> Account {
    Account(
        id: id, displayName: name, avatarData: nil,
        servers: [ServerInfo(id: host, name: host, url: URL(string: "https://\(host)")!)]
    )
}

#Preview("Chats – multi-account") {
    let accounts = [
        previewAccount("did:plc:real", "Real", "safe-haven.org"),
        previewAccount("did:plc:persona", "Persona", "pseudo.example"),
    ]
    return ChatsView().environmentObject(previewChatsState(accounts: accounts))
}

#Preview("Chats – multi-account (dark)") {
    let accounts = [
        previewAccount("did:plc:real", "Real", "safe-haven.org"),
        previewAccount("did:plc:persona", "Persona", "pseudo.example"),
    ]
    return ChatsView()
        .environmentObject(previewChatsState(accounts: accounts))
        .preferredColorScheme(.dark)
}

#Preview("Chats – single account") {
    let accounts = [previewAccount("did:plc:real", "Real", "safe-haven.org")]
    return ChatsView().environmentObject(previewChatsState(accounts: accounts))
}
#endif
